### ISSUE TITLE
Inline asm memory operands not properly referenced

### ISSUE THREAD

**Issue by N00byEdge** [ziglang/zig#10367]
### Zig compiler version

`0.9.0-dev.1952+5d1aab72d`

### Steps to reproduce

Equivalent C code, works fine with gcc:
```c
void f() {
    asm volatile("movl %[value], %[ptr]"
            :
            : [ptr] "m" (*(volatile int *)0x444444)
            , [value] "ir" ((int)0x414141)
            : "memory"
    );
}
```

In zig:
```zig
pub fn main() !void {
     asm volatile("movl %[value], %[ptr]"
            :
            : [ptr] "m" (@intToPtr(*volatile u32, 0x444444).*)
            , [value] "ir" (@as(u32, 0x414141))
            : "memory"
    );
}
```
The memory operand ("m") takes a memory reference, not a value (you should dereference a pointer or provide a value on the stack in the operand)

### Expected behaviour
Something equivalent to the C codegen:
```x86asm
 mov    eax,0x444444
 mov    DWORD PTR [rax],0x414141 // This is the generated inline asm instr
 ```

### Actual behaviour
Zig is deciding to dereference the pointer, but providing the memory location on the stack with the dereferenced value, not actually storing the value to the requested memory operand
```x86asm
 mov    eax,0x444444
 mov    eax, DWORD PTR [rax]
 // ...
 mov    DWORD PTR [rbp - offset], eax
 mov    DWORD PTR [rbp - offset], 0x414141 // This is the generated inline asm instr
 ```

### Notes
I know this may be a hard issue to solve, but I suggest at least disabling memory operands that take a pointer deref/other rvalue with a link to this issue for now. Passing a stack variable as a memory operand still works.

**Comment by andrewrk** [comment]
> I know this may be a hard issue to solve, but I suggest at least disabling memory operands that take a pointer deref/other rvalue with a link to this issue for now. Passing a stack variable as a memory operand still works.

I agree with your conclusion. I won't have time to do this before 0.9.0 on Monday but I would happily accept a pull request, and other than that it's in the 0.9.1 milestone.

Thanks for the report.

**Comment by N00byEdge** [comment]
Yeah I'm currently travelling so I won't have the time either, but if someone else wants to do it, my personal recommendation is a message that goes something like this `"This will miscompile: [link to issue] if the memory location is important. If you don't care about the location of the memory location (stack copy is fine), explicitly create a local copy of the value."`, or is that too verbose?

**Comment by andrewrk** [comment]
That's fine and actually pretty helpful and nice, for something that amounts to a compiler bug.

**Comment by andrewrk** [comment]
Confirmed still an issue in `0.11.0-dev.4271+1aacfa718`.

**Comment by N00byEdge** [comment]
I'm sorry @lockbox but your inline assembly doesn't make much sense. `"nextStage"` is not a valid input constraint, and you're not listing the registers you're clobbering. Not only that, but you're not using a memory reference to begin with.

**Comment by linusg** [comment]
https://codeberg.org/ziglang/zig/issues/31022
